### ISSUE TITLE
Push grouping to the fdw level. 

### ISSUE THREAD

**Issue by cristianprice** [Kozea/Multicorn#215]
Hi there,
Thanks for this awesome project. Quick question: is it possible to push the group by clause at the level of the foreign data wrapper? There are situations when i want to push this as low as possible to avoid network traffic by bringing data to postgres . 
Thanks,
C

**Comment by arturgspb** [comment]
Hi! Any updates?

**Comment by arturgspb** [comment]
@rdunklau ping, please

**Comment by rdunklau** [comment]
Hello !
In theory it is possible, however I have no idea how to map the API to do that cleanly.
We can keep this issue open in case someone has an idea or proposal, though.

**Comment by arturgspb** [comment]
Thx, a lot!

**Comment by gruuya** [comment]
Hey there! First of all, thanks for making Multicorn - we at [Splitgraph](https://www.splitgraph.com/) love it and have found it well suited for [our own needs](https://www.splitgraph.com/blog/foreign-data-wrappers).

In fact, we went a step further and just implemented aggregation/grouping pushdown capability into [our Multicorn fork](https://github.com/splitgraph/Multicorn). This capability is quite versatile, and basically allows for combinations of aggregation functions, `GROUP BY` clauses and qual (`WHERE`) clauses to be sent to the remote data source (if it supports those constructs of course). The relevant PRs are https://github.com/splitgraph/Multicorn/pull/1 and https://github.com/splitgraph/Multicorn/pull/2, while the practical [FDW example](https://github.com/splitgraph/postgres-elasticsearch-fdw/blob/master/pg_es_fdw/__init__.py#L118) that utilizes this feature (and only one so far) is our fork of Elastic search FDW. You can also take a look at our [suite of end-to-end tests](https://github.com/splitgraph/splitgraph/blob/master/test/splitgraph/commands/test_es_fdw_queries.py#L134-L726) that exercises the full range of these capabilities.

Basically, our implementation closely follows the standard FDW approach for handling these things, best illustrated by `postgres_fdw` extension itself.

That said, our fork has diverged from the upstream significantly, such that contributing these changes back seems quite laborious. Nonetheless, we figured drawing attention to this can't hurt, as perhaps someone may find it useful. :)

Cheers!

**Comment by df7cb** [comment]
Hi @gruuya, were you able to fix the problems with the recent PG major versions in that fork?

**Comment by gruuya** [comment]
Hello there @df7cb! We are (for now) focused on PG 12 exclusively, so we haven't dealt with any problems in the later PG versions.
